### PR TITLE
fetch data from EBI AlphaFold database

### PR DESCRIPTION
**Pull request by jonfunk21**
added functionality to download files from the EBI AlphaFold database.

PS. This is my first attempt to contribute to open-source projects. If anything is off, please tell me about it and I will fix it.

Best,
Jonathan

### CODE REVIEW

File: src/biotite/database/alphafold/__init__.py
Comment (padix-key) [inline src/biotite/database/alphafold/__init__.py:1]: This file needs to import the function from `alphafold.py` to be consistent with the typical Biotite import scheme in the form of
```python
import biotite.database.alphafold as alpafold`
alphafold.fetch(...)
```
Have a look at https://github.com/biotite-dev/biotite/blob/master/src/biotite/database/rcsb/__init__.py as template.

File: src/biotite/database/alphafold/download.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    database might return no or malformed data to you.
+    In these cases the request should be retried.
+    When the issue occurs repeatedly, the error is probably in your
+    input.
+
+    Examples
+    --------
+
+    >>> import os.path
+    >>> import requests
```
Comment (padix-key) [inline src/biotite/database/alphafold/download.py:61]: Unused import

```suggestion
```

File: src/biotite/database/alphafold/download.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    >>> file = fetch("P0DPA9", "cif", path_to_directory)
+    >>> print(os.path.basename(file))
+    P0DPA9-F1-model_v3..cif
```
Comment (padix-key) [inline src/biotite/database/alphafold/download.py:64]: I think the double suffix`..` result from a bug in your code. However, I did not find it from visual inspection. Does this appear in the current version of your code?

**Review by padix-key — CHANGES_REQUESTED** [review]
Thanks for implementing this! I think the Alphafold database support is a useful addition, especially as this topic already came up in a previous issue (#438). Since the PDB seemingly does not plan to directly store Alphafold structures in the near future, an interface to the EBI is reasonable in my opinion.

To test whether this implementation works properly some tests in `tests/database/test_alphafold.py` are necessary. You can take some inspiration from `tests/database/test_rcsb.py`, if you like. It also would be good to test the doctest string by adding the `alphafold` subpackage to `tests/test_doctest.py`.

Furthermore, I think it would be it would be nice, if not only *Uniprot IDs* would be recognized but also the IDs the RCSB PDB gives, for example `AF_AFP0DPA9F1`. This would allow downloading the IDs that were identified via `biotite.database.rcsb.search()`.

File: src/biotite/database/alphafold/download.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        else:
+            # 'file = None' -> store content in a file-like object
+            file = None
+
+        if file is None \
+                or not isfile(file) \
+                or getsize(file) == 0 \
+                or overwrite:
+            if format == "pdb":
+                r = requests.get(_standard_url + id + '-F1-model_v3' + ".pdb")
```
[the remark this answers is not in the document]
Comment (padix-key) [reply src/biotite/database/alphafold/download.py:100]: According to [this README](https://github.com/deepmind/alphafold/blob/main/afdb/README.md#format) currently all entries are provided as `v4`. Maybe we could create a global constant containing the version in this module? When we recognize that a new version exists, we just update this global variable.

If all entries are provided in the same version, we could even create a test function that increments this global version identifier and asserts that a sample file (any database entry) with the incremented version cannot be downloaded. If it is downloadable the test fails, as the global variable would not be up-to-date.

[the remark this answers is not in the document]
Comment (jonfunk21) [reply src/biotite/database/alphafold/download.py:100]: I like the idea of defining a global variable. Should the user have a choice of which version he wants to download, but if no preference is provided the latest version will be downloaded?

File: src/biotite/database/alphafold/download.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    """
+    # Structure file and FASTA file retrieval
+    # have different error messages
+    if any(err_msg in response_text for err_msg in [
+        "404 Not Found",
+        "<title>RCSB Protein Data Bank Error Page</title>",
+        "No fasta files were found.",
+        "No valid PDB IDs were submitted.",
+    ]):
+        raise RequestError("PDB ID {:} is invalid".format(pdb_id))
```
[the remark this answers is not in the document]
Comment (jonfunk21) [reply src/biotite/database/alphafold/download.py:141]: I'll adapt the section

File: src/biotite/database/alphafold/__init__.py
Comment (jonfunk21) [reply src/biotite/database/alphafold/__init__.py:1]: I'll add it

File: src/biotite/database/alphafold/download.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (padix-key) [reply src/biotite/database/alphafold/download.py:99]: This suggestion sounds good to me.

**Comment by padix-key** [comment]
Hi, are you still working on this? Maybe some other changes may be necessary now, since there is a new REST API for the AlphaFold DB that @BradyAJohnston shared in #438:

https://www.ebi.ac.uk/about/news/updates-from-data-resources/alphafold-database-ux-update/

https://alphafold.ebi.ac.uk/api-docs

I have not had a deeper look at it yet, but it may simplify/extend the access options to the database.

**Comment by padix-key** [comment]
I had a deeper look into the new API endpoints: They only provide additional information about the given UniProt entry but not the structure itself. So I think your current approach is still fine. However, one learning from the new API is that [BinaryCIF](https://github.com/molstar/BinaryCIF) seems now to be supported as well. This it not that relevant for now, but in the future Biotite might start support BinaryCIF as well.

**Comment by padix-key** [comment]
Hi, in #438 another user has shown interest to work on the AlphaFold DB interface. Hence, if you cancelled the development, I would like to give [dacarlin](https://github.com/dacarlin) the opportunity to work on this feature.

**Comment by padix-key** [comment]
Closed due to inactivity.

**Comment by jonfunk21** [comment]
Hey Patrick, sorry for not being responsive, I got too busy with other things. sure another user can work on the issue!

**Pull request closed without merging**